api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -0,0 +1,437 @@
+
+        obs_tensor = torch.Tensor(obs_encoded.to_numpy()).int()
+        if not self.sparse_X:
+            X_tensor = torch.Tensor(X.todense())
+        else:
+            coo = X.tocoo()
+
+            X_tensor = torch.sparse_coo_tensor(
+                # Note: The `np.array` seems unnecessary, but PyTorch warns bare array is "extremely slow"
+                indices=torch.Tensor(np.array([coo.row, coo.col])),
+                values=coo.data,
+                size=coo.shape,
+            )
+
+        if self.batch_size == 1:
+            X_tensor = X_tensor[0]
+            obs_tensor = obs_tensor[0]
+
+        return X_tensor, obs_tensor
+
+    def _read_partial_torch_batch(self, batch_size: int) -> Tuple[pd.DataFrame, sparse.csr_matrix]:
+        safe_batch_size = min(batch_size, len(self.obs_batch) - self.i)
+        slice_ = slice(self.i, self.i + safe_batch_size)
+        obs_rows = self.obs_batch.iloc[slice_]
+        X_csr_scipy = self.X_batch[slice_]
+        self.i += safe_batch_size
+        return obs_rows, X_csr_scipy
+
+    @property
+    # TODO: Retrieve next batch asynchronously, so it's available before the current batch's iteration ends
+    def obs_batch(self) -> pd.DataFrame:
+        """
+        Returns the current SOMA batch of obs rows.
+        If the current SOMA batch has been fully iterated, loads the next SOMA batch of both obs and X data and returns
+        the new obs batch (only).
+        Raises StopIteration if there are no more SOMA batches to retrieve.
+        """
+        if 0 <= self.i < len(self.obs_batch_):
+            return self.obs_batch_
+
+        pytorch_logger.debug("Retrieving next TileDB-SOMA batch...")
+        start_time = time()
+        # If no more batch to iterate through, raise StopIteration, as all iterators do when at end
+        obs_table = next(self.obs_tables_iter)
+        self.obs_batch_ = obs_table.to_pandas()
+        # handle case of empty result (first batch has 0 rows)
+        if len(self.obs_batch_) == 0:
+            raise StopIteration
+        self.X_batch = _fast_csr.read_scipy_csr(self.X, obs_table["soma_joinid"].combine_chunks(), self.var_joinids)
+        self.i = 0
+        self.stats.n_obs += self.X_batch.shape[0]
+        self.stats.nnz += self.X_batch.nnz
+        self.stats.elapsed += int(time() - start_time)
+        self.stats.n_soma_batches += 1
+        pytorch_logger.debug(f"Retrieved batch: shape={self.X_batch.shape}, cum_stats: {self.stats}")
+        return self.obs_batch_
+
+
+class ExperimentDataPipe(pipes.IterDataPipe[Dataset[ObsDatum]]):  # type: ignore
+    """
+    An iterable-style data pipe.
+    """
+
+    _query: Optional[soma.ExperimentAxisQuery]
+
+    _obs_joinids_partitioned: Optional[pa.Array]
+
+    _obs_and_x_iter: Optional[_ObsAndXIterator]
+
+    _encoders: Optional[Encoders]
+
+    _stats: Stats
+
+    def __init__(
+        self,
+        experiment: soma.Experiment,
+        ms_name: str,
```

Comment:
lets make the names consistent with the rest of the cellxgene_census package, e.g., `get_anndata` uses `measurement_name`, not `ms_name`

Reply:
To be clear:  I'm not suggesting you change the signature on this method/class.  Rather that a "sugar" factory function may be required.  this is entirely @pablo-gar 's call (and could wait for a future PR IMHO).

Reply:
got it! update the TODO to reflect